### ISSUE TITLE
[Bug]: missing integrity checks let attacker forge arbitrary message content using a known plaintext attack and replaying messages even when PSK is not known

### ISSUE THREAD

**Issue by Jorropo** [meshtastic/firmware#4030]
### Category

Other

### Hardware

Not Applicable

### Firmware Version

all ¿ (fundamental protocol issue)

### Description

This was discussed with @caveman99 in `#contributor-lounge` he gave me the go ahead to post it here.

Here is pseudo code targeting the admin module, this would let an attacker observe an admin packet, modify it and change more or less any settings at will on all the nodes using this admin PSK:
```py
def encryptPacket(nodeid, packetid, key, message):
 bitstream = generateAesCtrBitstream(uint64bytes(packetid) + uint32Bytes(nodeid) + uint32bytes(0), key, len(message))
 return bitstream ^ message

decryptPacket = encryptPacket

setOwnerPacket = createSetOwnerPacket(to, newName) # this assume this has predictable output, the attack is harder if this does not, but not impossible, the main drawback is that it would require many round trip with modified packets to guess the order of fields
packetId = generatePacketId()
cypherText = encryptPacket(sourceNodeId, packetId, key, setOwner)
sendLora(sourceNodeId, packetId, cypherText)

# let's assume I am a an attacker, all of the following code run on my computer, I just observed some packet I couldn't decrypt it, and I then see a new nodeinfo with a changed name
# all of this to say, let's assume I was able to learn `to` and `newName` from contextual clues
# I also know `sourceNodeId`, `packetId` and `cypherText` since theses are sent in clear text
setOwnerPacket = createSetOwnerPacket(to, newName) # recreate the plaintext, there are many other ways to achieve this even when the whole content cannot be guessed by doing trial and error
bitstream = cypherText ^ setOwnerPacket # the whole magic is happening here
# I can now forge abitrary content that is len(bitstream) long
maliciousPayload = createSetChannelPSK(to, index=1, psk=PSKControlledByAttacker) # assume this is setting admin channel psk
cypherText = bitstream[:len(maliciousPayload)] ^ maliciousPayload # encrypt it back
sendLora(sourceNodeId, packetId, cypherText) # replay spoofed packet with modified content
# you might need to send other packets first to make the node forget it has seen this packet else it might disregard it as duplicated
# you could also jam the original packet so it never reach the target node
```

---

The fix I recommend is to use AEAD.
I have prior experience with AES-GCM it is AES-CTR (what we are currently using) with a bit of extra math, this adds a 16 bytes hash after the message, it is accelerated on ESP32 and the extra math is not extremely expensive.
The main drawbacks:
1. The biggest admin packets don't have enough room for a 16 bytes auth.
  This is not a very hard engineering problem, protobuf isn't that efficient and I guess we could find a way to gain a few bytes.
2. Backward compat / extra 16 bytes cost on all packets.
  This can be solved by making this optional when configuring your channel, most channels like `AQ==` would gain nothing but we could enforce this for `admin`. So users would be able to toggle if they want integrity protection 
3. nrf52 chips don't support AES-GCM acceleration, they do support AES-CCM tho, which also do AEAD, I am not familiar as familiar with AES-CCM but my limited opinion is that it should be good enough, it is based on AES-CBC instead of AES-CTR which GCM is based on.
  ESP32 chips also support AES-CCM acceleration.
  So **AES-CCM should be a good alternative**, drawback 1 and 2 still apply.

AES-CCM is often what is used in WPA3.
AES-GCM is used in an overwhelming of TLS1.3 connections (altho Chacha20-Poly1305 is also often implemented as a fallback for CPUs who lack AES hardware acceleration), updated TLS1.2 clients also often prefer AES-GCM based suites. You are extremely likely to be using AES-GCM to view this very same issue.

This would not solve all the issue present in meshtastic's encryption, this only solve Integrity but not Perfect-Forward-Secrecy or Authentication.

### Relevant log output

```Shell
N/A
```

# Is it dangerous to use the admin module ?

Probably not:
The admin module is most useful on unattended nodes where a physical attack is significantly easier and faster to perform.
This require to first capture a valid packet sent on the admin channel which is unlikely because how often do you change configuration settings ?
The attacker then need to identify which packet were sent on the admin channel, this is easier said than done, packets do not publicly indicate which channel they are a part of, there is no easy way to differentiate a packet on the admin channel to any random encrypted packet.
Then the attacker must correctly guess some bits in the admin message, in general each bit guessed correctly can be changed to attacker's value of choice. Admin messages a fair bit complex.

Examples of things that are "easy":
> You use the admin module to change the role to ROUTER_CLIENT.
> An attacker can then change the role CLIENT_MUTE.

This is because a huge portion of the message is identical and the device role is broadcasted in nodeinfo making guessing easier.
Changing the target of an admin message that clearly reflect changes in public info is also "easy".
For example you have two nodes using the same admin channel, you update some configs on node A, an attacker can then send the same configuration to node B. The problem with this is that the attacker does not know which nodes share the same admin channel.

**Tl;Dr: the vast majority of attacks are harder and it take significantly more time than gaining physical access to your node.**
Most impacts of such attack would be making the mesh unreliable, using the same hardware this can be done by setting HOP_LIMIT=7, override the duty cycle checks and configure the node to spam random messages every second completely hogging the spectrum.

## Workarounds:

- Do not use the remote administration module, BLE and USB can do the job sometime.
- For nodes relying private channels ensure all channels your node is part of as private and secure, this will make extremely hard to guess valid values since all packets will look like encrypted garbage.
- Use different admin channel PSKs, this make it impossible to redirect config packets destined to one node, to an other node.

**Comment by Jorropo** [comment]
If implementing AEAD is not possible, a correct MAC (which include both message content and PSK) is also acceptable, this historically created Side-Channel-Attacks in TLS1.2 which required TLS1.2 implementations to add mitigations and TLS1.3 removed all non AEAD support.

Even in the worst correct case MAC with SCA is way better than the current state of things.

**Comment by Jorropo** [comment]
Actually I over-complicated padding.
Message length is not encrypted (it is stored in the explicit LoRa header), you can simply truncate the message to the length you want.

**Comment by xnyhps** [comment]
> this adds a 16 bytes hash after the message

16 bytes is ideal, but shorter authentication tags can be used for AES-GCM/AES-CCM. 12 bytes would still be fine, and depending on the situation, even 8 bytes can be used. The slow rate at which messages can be sent could justify using shorter tags.

**Comment by xloem** [comment]
If the fundamental protocol will change, given issues like this have been so extant, I would propose finding existing cryptographic solutions and systems to use as much as possible, so as to share work in maintaining them with others. There is AI-driven attack research now.

**Comment by EternityForest** [comment]
#2610 is possibly relevant.

As I understand it, MACs are *not* subject to the birthday paradox (Reference: https://crypto.stackexchange.com/questions/50085/are-macs-vulnerable-to-birthday-attacks).

An 8-byte MAC, at 500 messages per second, would take hundreds of thousands of years to get one fake message through, if I understand the math: 2^63÷(500×3600×24×365)

A four-byte MAC at a still unrealistically high 10 messages per second would take 6 years. But, you could lock out for a few seconds if you get too many incorrect MACs, so it would be more like hundreds of years.


I think a more relevant attack would be replay attacks.  Not likely for the admin channel, but very likely in some use cases.


I've mentioned this before in other threads, but rather than an 8-byte MAC, what about a 4 or five byte MAC, with a timestamp?



Setting the time could be done with a dedicated time sync packet, itself authenticated.  If you don't get a time update in a few months, fall back to ignoring time.


The timestamp could actually just be implicit, you could make the MAC key a hash of (Timestamp+actual key).   Then you only need the four extra bytes.  

The timestamp would just have to only be precise to a few minutes, and you'd have to try decrypting with both possible keys, if you're right on the edge of a time boundary.


Or, you could have an explicit timestamp, in the plaintext associated data, which would incidentally also serve as a time sync packet, that any other node could also read, just without the authentication.



For replay attacks *within* few minute time window, a simple one byte message counter counter be used, that resets at the start of a time window.  Just keep track of the time you last got each of the possible values.

It breaks down if you have multiple senders to one receiver within a time window, but that's fine for admin channel,  and most home automation tasks, and could be disabled for the rest.

If nodes randomize their starting counter values it would mostly work with multiple nodes anyway.

With more memory you could also just keep track of all messages you got in the last few minutes.

**Comment by Jorropo** [comment]
Ok so I was implementing CCM-optin and https://github.com/rweather/arduinolibs/issues/52 is a blocker for linux-native and rp2040 targets. Chances are I'll fix this upstream.

**Comment by fifieldt** [comment]
Hi @Jorropo . In 2.5.0 we've re-done the admin channel with some PKI , which should fix this issue.

https://www.youtube.com/watch?v=7q-dNZzpUlM

**Comment by xnyhps** [comment]
I'm a bit confused why this issue is reported only for the admin channel, and (as it appears currently) only addressed for the admin channel. Any private channel would be affected by this in the same way. Of course, the admin channel is the most tempting as it allows re-configuring a node, but even on other channels it would be an issue.

I could spin up a repeater that always flips a few bits before forwarding a message. Any receivers who got the message via this repeater won't be able to determine that the message was manipulated (unless it corrupts the protobuf structure, but I can probably guess where the actual meaningful part is). As repeaters are invisible, I think it wouldn't even be possible for the nodes around me to ignore all messages from my repeater.

Is this correct or am I missing something?

**Comment by Jorropo** [comment]
@xnyhps this has nothing to do with admin channel and affects everything as you correctly pointed out.

The idea I had is that if you randomly flipped bits there would presumably be a carbon based intelligence that would understand a glitchy hopefully not valid UTF8 message to be something wrong while a silicon based intelligence wouldn't.

**Comment by Jorropo** [comment]
I've tried fixing it by adding an AEAD channel mode option but changing the architecture of the C++ code to integrate an extra option, do it correctly on all targets, was not fun so I gave up.
If someone care about fixing this it is worth reopening, I don't really.

**Comment by pqcfox** [comment]
Hi! I'm personally very interested in implementing this as a feature, and can take on the workload--as an applied cryptographer having an AEAD channel mode would be important to me.

Would it be possible to reopen this issue, just so the work I'm doing is going toward an existing issue? Thanks!

**Comment by garthvh** [comment]
@pqcfox done good luck!

**Comment by david415** [comment]
> [@pqcfox](https://github.com/pqcfox) done good luck!

don't be an ass

**Comment by david415** [comment]
> Hi! I'm personally very interested in implementing this as a feature, and can take on the workload--as an applied cryptographer having an AEAD channel mode would be important to me.
> 
> Would it be possible to reopen this issue, just so the work I'm doing is going toward an existing issue? Thanks!

Food for thought:

I am so embarrassed for Meshtastic because there are many MANY definitions of a bad person, a person who causes harm to others or who through their inaction allows harm to come to others, however in today's day an age there is also a kind of cryptographic / information theoretic definitions of a "bad person" which certainly is a person who releases insecure cryptographic protocols that could allow for human rights violations.

This is not a game. People are dying in the streets. The secret police in my country are acquiring more and more power. The people need strong encryption.

**Comment by randshell** [comment]
Hi @david415,

I see your point of view, as I am big fan of privacy as well, though I can't speak for the attacks on human rights that you're living with there.

Meshtastic, as of now, has quite a few limitations when it comes to privacy and security. In fact, it was born as a hobby project and, while it has greatly improved since then, it's in no way recommended for activism or any other high-risk scenarios. In their documentation there's also no mention that this project is fit for these use cases.

If I can suggest, ideally you'd go on projects that are centered around what you need and made specifically for that purpose. Since I've bumped into some tools and apps that were nicely marketed, perhaps with the bonus that they're open source, I feel the need to mention to always do your own research. Check their claims on other websites. Check if they have independent security audits to back up what they're saying, read these audits too. Check if they have any public CVE's or other security advisory published.

It happened at least once for me to check out a privacy-focused decentralized VPN, which I won't mention, that didn't really have any audits or the documents of these analysis weren't available to download anywhere. I'd be weary of such projects.


Best of luck,

**Comment by pqcfox** [comment]
As an update, after review of the Meshtastic codebase, I don't feel confident that any contribution I make in a reasonable amount of time would be secure to the extent that I'd be comfortable having my name on it as a security feature. I'll sadly not be able to deliver on this feature, but I'm interested to watch and see if anyone else does.

There's simply too many interdependent moving parts involved in routing for me to be confident that anything I write wouldn't introduce an additional vulnerability or compromise the security of the AEAD channel in some way, and given the underlying cryptographic primitive implementations for some platforms are presently unaudited, I don't feel as though anything I add could be considered strictly secure to an extent that I'd be comfortable guaranteeing without extensive further work myself, likely beyond my capacity.

That all said, I do think Meshtastic is a fantastic project and one that I'm thrilled has had such remarkable traction over the last year, and I'm really excited to see all the amazing things it's been doing for mesh networking as the project grows! Thanks for your time, and my apologies for any inconvenience.

**Comment by nathanpegram** [comment]
Does this issue imply a vulnerability to replay that could be used to control devices with the remote hardware module configured?

**Comment by pqcfox** [comment]
IIRC that uses the PKI infra, so it shouldn't be affected by this issue in particular (?)

**Comment by nathanpegram** [comment]
Reading the [documentation](https://meshtastic.org/docs/configuration/module/remote-hardware/ ) for how to use it makes it look like it just uses a channel with a PSK. But I am admittedly very new to meshtastic so I could be misunderstanding things.

**Comment by khimaros** [comment]
is this issue resolved?

**Comment by Jorropo** [comment]
No, the bot is annoying here.

**Comment by matutetandil** [comment]
Hi all — I've been analyzing this issue and I'd like to offer to take a crack at a PR.

After reading through the firmware codebase, I noticed that AES-CCM is already implemented and working for PKI-encrypted DMs (`CryptoEngine::encryptCurve25519` / `decryptCurve25519` using `aes_ccm_ae` / `aes_ccm_ad`). The core AEAD primitive is already there and battle-tested — it's just not used for PSK channel encryption, which still goes through the unauthenticated AES-CTR path.

The approach I'm considering:

1. **Add an opt-in AEAD mode for PSK channels** — a new field in `ChannelSettings` (e.g., `bool use_aead = N`) that switches the channel from AES-CTR to AES-CCM
2. **Reuse the existing `aes-ccm.cpp` implementation** — no new crypto libraries needed, just route PSK channel packets through the CCM path instead of CTR
3. **8-byte auth tag** — fits within the existing packet budget (~8 bytes overhead, leaving ~225 bytes of payload). As @EternityForest noted, at Meshtastic's message rate an 8-byte tag gives hundreds of thousands of years before a brute-force forgery
4. **Backwards compatible** — `use_aead = false` (default) preserves current AES-CTR behavior. Nodes on the same channel must all opt in. The flag travels in the ChannelSet URL/QR, so sharing "just works"
5. **On receive, try CCM first (if enabled), fall back to CTR** — allows gradual migration within a mesh

I'm aware that @Jorropo and @pqcfox ran into complexity integrating this across all targets. I'll focus on keeping the change as surgical as possible — ideally just modifying the encrypt/decrypt dispatch in `Router.cpp` and `Channels.cpp` without restructuring the crypto architecture.

I'd also need a small protobuf change in `meshtastic/protobufs` for the new `ChannelSettings` field.

Before I start coding — does this approach sound reasonable to the maintainers? Any concerns or preferences on the implementation strategy? Happy to discuss further.

**Comment by pqcfox** [comment]
Hiya! Thanks for the tag ^^

I would suggest that this potentially could use some careful modeling out-of-band. In particular, from my past review of this issue, here's some relevant concerns:

- The use of an 8-byte auth tag lowers collision resistance to 32 bits, which depending on the use case may not be advisable and could warrant a review of the protocol as a whole

- User-configurable cryptography not aimed at cryptographic engineers can often end in disaster; the AEAD option could be suited as a one-time feature enable flag, but having many variants of a protocol in flight makes analysis and later protocol formalization much harder

- In particular, downgrading attacks might be relevant if receiving CTR messages causes the rest of an exchange to be unauthed

- The CCM implementation should probably be run through Wycheproof, as it's (to my knowledge) not presently audited or tested for correctness (though IIRC the external crypto libs used aren't audited either)

In particular, I do think as a matter of shared responsibility that any additional claim of cryptographic guarantees (e.g. authenticated channels) needs to be paired with a lot of attention to these kinds of underlying details which might undermine the claimed property. 

(This of course can be noted as experimental work, but I'd be remiss if I didn't wave my arms a bit and say that there are possible risks which need to be mitigated before one can responsibly claim to users that they can rely on a new property; some of these are less concerns regarding the suggested strategy as much as what I understand to be proximate needs that could use attention in conjunction with this issue.)

**Comment by matutetandil** [comment]
Thanks @pqcfox for the detailed feedback — exactly the kind of review this needs. Let me address each point:

**On the 8-byte tag / 32-bit collision resistance:**

Fair point on the birthday bound. While forgery resistance at 2^64 is more than sufficient at LoRa message rates, I agree that formal protocol analysis should account for the birthday bound. A **12-byte tag** seems like the sweet spot — gives 2^48 birthday resistance while only costing 4 extra bytes over the 8-byte variant (225 → 221 bytes usable payload, still plenty for text messages and telemetry). The existing PKI mode uses 8-byte tags, but for a new channel mode we have the opportunity to be more conservative from the start.

**On user-configurable cryptography:**

Fully agree. My intent is a simple one-way flag — once AEAD is enabled on a channel, there's no "choose your cipher" UI. Think of it as a protocol version bit rather than a crypto menu. Ideally it becomes the default in a future firmware version once adoption is sufficient.

**On downgrade attacks:**

This is the most important point. I now think the right approach is: if AEAD is enabled on a channel, **reject CTR-only packets entirely** — no fallback. This makes the security property binary: either your channel is authenticated or it isn't. A mixed-mode fallback would undermine the entire purpose. The trade-off is that all nodes on a channel must upgrade together, but that's enforceable through the channel sharing URL (the AEAD flag travels with the ChannelSet).

**On CCM implementation testing:**

Agreed — I'll include Wycheproof test vectors for the CCM path in the PR. The existing `test/test_crypto/test_main.cpp` already has some CCM coverage, but comprehensive test vectors would raise confidence.

**On responsible claims:**

100% agree. I'd suggest marking this as experimental initially — documented as "channel authentication (experimental)" with clear language about what it does and doesn't guarantee. No overclaiming.

Would this adjusted approach address your concerns? Happy to iterate further before writing code.

**Comment by pqcfox** [comment]
@matutetandil absolutely! Kudos for the very thoughtful attention to detail here, I think this makes a lot of sense. Your planned approach on each point seems more than reasonable to me for putting together experimental support (and thank you for the correction on the CCM functests!).

(I do think the larger issue regarding public messaging to users with increasingly disparate threat models re: how much they can rely on claimed properties is something that does need separate attention in public materials, but by no means should be blocking for this work. ^^)

**Comment by matutetandil** [comment]
## Implementation PRs submitted

I've submitted two PRs that implement AES-CCM authenticated encryption for PSK channels, as proposed in this issue:

### PR 1: Protobufs — [meshtastic/protobufs#868](https://github.com/meshtastic/protobufs/pull/868)
- Adds `bool use_aead = 8` field to `ChannelSettings` in `channel.proto`
- Per-channel opt-in flag (default `false`, preserving existing AES-CTR behavior)

### PR 2: Firmware — [meshtastic/firmware#9749](https://github.com/meshtastic/firmware/pull/9749)
- **AES-CCM encrypt/decrypt** for PSK channels (`encryptPacketCCM` / `decryptPacketCCM` in `CryptoEngine`)
- **12-byte authentication tag** — prevents message forgery, bit-flipping, and the known-plaintext attack described in this issue
- **No CTR fallback** on AEAD-enabled channels — tampered or unauthenticated packets are rejected
- **Hash differentiation** so AEAD and non-AEAD channels with the same PSK route independently
- **Backward compatible** — existing channels continue using AES-CTR unless `use_aead` is explicitly enabled
- **10 unit tests** including round-trip (AES-128/256), tampered ciphertext, tampered tag, wrong PSK, wrong sender, and sweep tamper tests

The AES-CCM primitives already existed in the codebase for PKI direct messages — the implementation reuses them and removes the `#if !MESHTASTIC_EXCLUDE_PKI` guards so they're available unconditionally. Key promotion (16-byte PSK → 32-byte for AES) is handled transparently.

### Bonus fix: stack buffer overflow in `aes_ccm_encr`
While testing under AddressSanitizer (CI `coverage` environment), I found a pre-existing buffer overflow in `aes_ccm_encr()`: the last partial block wrote a full 16-byte AES output directly to the caller's buffer before XOR, overflowing when the buffer was exactly sized to the plaintext. Fixed by using a temporary buffer for the partial block, matching the pattern already used in `aes_ccm_encr_auth()` and `aes_ccm_decr_auth()`.

### Also fixed: portduino `-s` flag preventing config YAML loading
The `-s` (simradio) flag in `PortduinoGlue.cpp` used an if/else-if chain that short-circuited YAML config loading. Separated config loading from the simradio override so both work together (relevant for simulator testing with `EnableUDP`).

**Comment by Avamander** [comment]
I think this proposal needs careful review before anything is implemented. There are quite a few pitfalls with CCM usage in general and in implementing it. Rolling out a cryptosystem, especially something for messaging is _very_ difficult to do correctly. Mistakes are even more annoying to remedy.

**Comment by postlmc** [comment]
We validated this end-to-end against current firmware while building tooling for a DEF CON 34 Ham Radio Village talk, and want
to confirm it in support of the fix.

Channel messages are AES-CTR with no MAC, so the keystream for a given (packet_id, from_node) nonce is recoverable, from the PSK
or from a single known-plaintext without the PSK, and a forged ciphertext can be reinjected under the same nonce. Recipient nodes
decrypt and display the forged content as though it came from the original sender. Our forge.py runs the whole loop against a live
node: capture, recover the keystream, reinject.

One composition we would flag: pairing forgery with packet dropping lets an attacker spoof a routing ACK, source-spoofed to the
recipient and referencing the dropped packet id, so the sender sees a false delivery confirmation for a message that was never
received. We have reported that separately via the Security tab. The fix in this issue, authenticated integrity that binds source
identity, closes both.

Happy to share the reproduction scripts and the firmware patches privately if useful (our firmware fork is not public yet).

**Comment by EternityForest** [comment]
@matutetandil Does your implementation do anything against replay attacks?

I might be repeating myself here, but I still think that's important.

As I understand it MACs are NOT subject to the birthday paradox, because it does not matter if your message collides with some random other messages from 5 years ago, unlike an IV collision where it very much IS subject to the birthday paradox, because any collision is an IV reuse.


A 48 or 64 bit MAC combined with a 32-bit timestamp would give both replay protection AND extra IV reuse security(although IV reuse is already unlikely) while still providing hundreds of years of brute force resistance even at a very unrealistic message rate without any rate limits.

A real implementation has no reason to allow more than a few bad packets per second.

**Comment by matutetandil** [comment]
Status update on the AEAD implementation, plus an answer to @EternityForest that I owed from July.

## Replay: no, and deliberately so

@EternityForest — straight answer: **the PR does nothing about replay.** AEAD buys integrity and authenticity, not freshness. A captured valid packet, replayed unchanged, still authenticates, because it *is* authentic. What it kills is the attack this issue is actually about: recovering keystream from a known plaintext and reinjecting forged content under the same nonce. Forgery, not repetition.

The only thing standing against replay today is `PacketHistory::wasSeenRecently()`, a bounded recent-packets table keyed by `(sender, id)`. That suppresses a replay while the record is still resident and nothing at all once it has been evicted. It is duplicate suppression that happens to inconvenience an attacker, not a security control.

Your MAC-plus-timestamp sketch is the right shape for fixing that, and I agree the birthday-bound argument does not apply to a MAC the way it applies to an IV. But it needs things this PR does not: a timestamp on the wire, a replay window, and a policy for nodes whose clocks are wrong or absent — which on this hardware is a lot of them. That is a second wire-format change with its own failure modes, and bolting it onto a PR that has been open since February would sink both.

So my position is that these are two changes, in this order. AEAD first, since it is the smaller change and the one that closes the reported attack. Replay protection second, on top of it, with the timestamp designed properly rather than smuggled in. Happy to work on the second one once the first lands — but I would rather not hold the forgery fix hostage to it.

@postlmc — your independent confirmation is useful, particularly the ACK-spoofing composition. That one is worth separating out: the forged-ACK half is closed by authentication, since the attacker can no longer mint a routing packet that verifies under the channel key. The packet-dropping half is not, and no amount of AEAD will fix it — an attacker who can suppress transmissions can always deny delivery. Authentication turns a false *positive* delivery confirmation into a missing one, which is the honest failure mode but still a failure.

## Where the implementation stands

@Avamander, your caution from March was fair, so here is what the code has actually been put through since:

- **Known-answer tests against RFC 3610 §8** — Packet Vectors #1, #2 and #7, covering M=8 and M=10, each ending in a partial block, plus AAD-corruption rejection. The CCM primitive is now pinned to published data rather than to its own output.
- **`test_crypto` is 15/15 under AddressSanitizer.** The AEAD block alone covers 12 scenarios: round-trip on AES-128 and AES-256, tampered ciphertext swept byte by byte including the tag, wrong PSK, wrong sender, undersized packet, empty PSK, and the `-1` invalid-key sentinel.
- **Review has been substantive**, and I have changed the design twice because of it. @robekl caught an empty-PSK crash and an ignored return value. @Jorropo pointed out that promoting a 16-byte PSK to AES-256 bought two extra rounds and no security, so 16-byte PSKs now use true AES-128. @jp-bennett has asked for the authentication to cover the sender and destination IDs, which I am working on now — `to` is currently bound to nothing, so the destination in the header can be rewritten with the tag still verifying.

Two pre-existing bugs fell out of the testing and have been split into their own PRs:

- **[#11347](https://github.com/meshtastic/firmware/pull/11347)** — out-of-bounds write in `aes_ccm_encr()` on a trailing partial block. Latent on `develop`, since every current caller passes a buffer with slack, but the decrypt path clears it by only a few bytes. **Approved today.**
- **[#11348](https://github.com/meshtastic/firmware/pull/11348)** — `-s` short-circuiting config loading in portduino. Under discussion; @jp-bennett reads the current behaviour as intended, and the fix may end up narrower.

## The actual blocker

[**meshtastic/protobufs#868**](https://github.com/meshtastic/protobufs/pull/868) — one field, nine lines:

```proto
bool use_aead = 8;
```

Open since **25 February**, `MERGEABLE`, still with zero reviews. Field 8 of `ChannelSettings` is free on `master` and nothing merged since has taken it.

Everything above is downstream of that. Without the field the firmware does not compile, so [#9749](https://github.com/meshtastic/firmware/pull/9749) currently carries a hand-written `channel.pb.h` — which is precisely what @caveman99 asked me to drop, and which I cannot drop until `update_protobufs.yml` can regenerate it from a merged protobuf. That workflow is `workflow_dispatch`, so it is a maintainer action, not something I can run.

@thebentern @jamesarich @caveman99 — a review on protobufs#868 unblocks the whole chain. If the field number, name, or default is wrong, say so and I will change it; if the answer is that AEAD should not be per-channel at all, that is worth knowing before more review time goes into the firmware side. Right now the work is stalled on nine lines that nobody has looked at in five months.

**Comment by yulinlina** [comment]
**Approach:** Add a security documentation note describing the PSK integrity limitation, attack impact, and required authenticated-encryption fix.

**Commit message:** `docs: document PSK integrity limitation`

**Changes:**

- `docs/security/psk-integrity-limitation.md` (create)

**Comment by caveman99** [comment]
@matutetandil My Main gripe with this PR is, that i refuse to wade through a wall of AI storytelling to extract the impact this has. There's a human at the other end of this conversaation.

**Comment by yulinlina** [comment]
Fair point. I over-explained. Here is the short human version:

**What is broken today**  
Meshtastic encrypted packets use AES-CTR without integrity protection. If an attacker captures a packet and knows/guesses some plaintext, they can recover keystream for that `(from, packet_id)` nonce and construct modified ciphertext that still decrypts as a packet from the original sender — without knowing the PSK. For admin packets, that can mean unauthorized configuration changes; for normal messages, impersonation.

**What the AEAD change fixes**  
It adds authentication and binds the relevant packet metadata into the authenticated data. If payload or identity fields are altered, the MAC fails and the packet is dropped. This blocks the known-plaintext/keystream forgery described in this issue, assuming the PSK remains secret.

**What it does not fix**  
Replay. A captured valid packet can still be resent. `PacketHistory::wasSeenRecently()` is only duplicate suppression, not a real replay policy. Freshness/replay defense needs a separate mechanism, likely involving timestamps/sequence numbers and a replay window, with special handling for clockless nodes.

Suggested concise PR description:

```markdown
Security impact: replaces unauthenticated AES-CTR channel encryption with AEAD and authenticates packet identity fields. Prevents forged or modified packets created via known-plaintext keystream recovery. Does not provide replay protection; replay defense requires a separate freshness/replay-window mechanism.
```

If this wording is acceptable, I can trim the description down to this level and keep replay as a separate follow-up.

**Comment by garthvh** [comment]
Adding a set of concrete 2.8-era targets for this same primitive, from the "meshtastic-is-dead" talk (Crypto attacks, "Tamper without the key").

The root cause is the one this issue already describes, AES-CTR with no MAC, and the title already covers the key point that the PSK isn't needed. What the talk adds is that for certain fields you don't need to reconstruct the plaintext at all. This issue's worked example builds `bitstream = cypherText ^ knownPlaintext` to forge chosen content, which needs a good guess at the original message. But a single bit flipped at a known offset has a deterministic effect on the plaintext regardless of what the rest of the message says, and several fields worth attacking now sit at predictable offsets:

- **`bitfield`**, at a fixed offset from the end of the encoded `Data`. Flipping one bit sets `want_response` (every receiver that decrypts it generates a reply, an amplification lever) or `OK_TO_MQTT`. The latter is the interesting one: it pushes traffic to public MQTT against the sender's explicit choice, so an attacker with no key for the channel can cause its contents to be uplinked to the internet where a gateway that does hold the key will publish them. The shifts are public constants, `BITFIELD_OK_TO_MQTT_SHIFT` and `BITFIELD_WANT_RESPONSE_SHIFT`, applied in `perhapsEncode()` (`src/mesh/Router.cpp`).

- **NodeInfo `public_key`**, 32 bytes near the end of the message. XORing garbage over it needs no knowledge of the original and corrupts the key a receiver binds to that node number under TOFU, a targeted denial of PKI messaging with that peer. Substituting a chosen key requires knowing the original bytes, but those are broadcast in the clear in other NodeInfo packets, so that precondition is often satisfiable.

- **`id`** is a plaintext header field, not ciphertext, so rewriting it needs no crypto trick at all. This issue notes in passing that a replay may be dropped as a duplicate; rewriting `id` is the general answer to that, since dedup keys on it.

None of these are separate bugs. They're the same missing-integrity root cause this issue was opened for, and they're consistent with what `SECURITY.md` documents as accepted. Recording them here because the accepted-limitation language is written around *possession of the PSK*, and these particular edits sit outside that framing: the tamper primitive needs no key, only a known offset. Worth weighing when the AEAD discussion above gets revisited.

2.8's XEdDSA signing mitigates part of this for broadcast traffic, since a receiver that has seen a node sign can reject unsigned packets claiming that identity, and a bit-flipped packet fails signature verification. It does not cover unicast or unsigned-capable paths.

**Comment by matutetandil** [comment]
@garthvh - AEAD landed in develop this morning as d05fbec64 (#9749), so on channels with `use_aead` set the primitive your first two targets rely on is closed.

The `bitfield` and the NodeInfo `public_key` both sit inside the encoded `Data`, which is what gets encrypted, so the CCM tag covers them: a bit flipped anywhere in the ciphertext fails authentication and the packet is dropped instead of decoded. `from` and `to` are passed as associated data as well, so the destination can't be rewritten either. The hop fields are deliberately excluded, since relays legitimately rewrite them.

Your third point still stands. `id` is a plaintext header field rather than ciphertext, so nothing here touches it. That's the same thread as the Store & Forward problem @caveman99 wrote up on the PR: `preparePayload()` preserves `from` and `id` while rewriting `to`, so the replay reuses (key, nonce). The fix is a fresh id for the replay, but the phone apps dedup and thread S&F history on that id, so it needs its own PR with app-side coordination.

Worth being clear that AEAD is opt-in and off by default, so the default configuration is unchanged and everything this issue documents still applies there.
